e3nn/kernel_mod.py

```diff
@@ -0,0 +1,118 @@
+# pylint: disable=missing-docstring, line-too-long, invalid-name, arguments-differ, no-member
+import math
+import torch
+
+import e3nn.SO3 as SO3
+
+
+class Kernel(torch.nn.Module):
+    def __init__(self, Rs_in, Rs_out, RadialModel, get_l_filters=SO3.selection_rule, sh=SO3.spherical_harmonics_xyz, normalization='norm'):
+        '''
+        :param Rs_in: list of triplet (multiplicity, representation order, parity)
+        :param Rs_out: list of triplet (multiplicity, representation order, parity)
+        :param RadialModel: Class(d), trainable model: R -> R^d
+        :param get_l_filters: function of signature (l_in, l_out) -> [l_filter]
+        :param sh: spherical harmonics function of signature ([l_filter], xyz[..., 3]) -> Y[m, ...]
+        :param normalization: either 'norm' or 'component'
+        representation order = nonnegative integer
+        parity = 0 (no parity), 1 (even), -1 (odd)
+        '''
+        super().__init__()
+
+        self.Rs_in = SO3.simplifyRs(Rs_in)
+        self.Rs_out = SO3.simplifyRs(Rs_out)
+
+        def filters_with_parity(l_in, p_in, l_out, p_out):
+            nonlocal get_l_filters
+            return [l for l in get_l_filters(l_in, l_out) if p_out == 0 or p_in * (-1) ** l == p_out]
+
+        self.get_l_filters = filters_with_parity
+        self.sh = sh
+
+        assert isinstance(normalization, str), "normalization should be passed as a string value"
+        assert normalization in ['norm', 'component'], "normalization needs to be 'norm' or 'component'"
+        self.normalization = normalization
+
+        def lm_normalization(l_out, l_in):
+            # put 2l_in+1 to keep the norm of the m vector constant
+            # put 2l_ou+1 to keep the variance of each m component constant
+            # sum_m Y_m^2 = (2l+1)/(4pi)  and  norm(Q) = 1  implies that norm(QY) = sqrt(1/4pi)
+            lm_norm = None
+            if normalization == 'norm':
+                lm_norm = math.sqrt(2 * l_in + 1) * math.sqrt(4 * math.pi)
+            elif normalization == 'component':
+                lm_norm = math.sqrt(2 * l_out + 1) * math.sqrt(4 * math.pi)
+            return lm_norm
+
+        # Clebsch-Gordan for filter, input, output
+        # Rs_filter contains all degrees of freedom and L's for filters
+        # paths contains [l_in, l_out, l_filter for each filter channel]
+        Rs_filter, filter_clebsch_gordan, paths = SO3.tensor_productRs(Rs_in,
```

Comment:
@blondegeek Can you prove/argue why this is the correct function to use? (Since we are not doing the tensor product of `Rs_in` with `Rs_out`)

Reply:
Am I correct in interpreting your question to be, why should we include the Clebsch-Gordan coefficients in `Kernel` when they are really needed for the `Convolution`? If that is so, I completely agree. And it's something we should consider perhaps storing in `Kernel` (since it's created when we create `Rs_filter`), but not applying until the `Convolution`.

Since, it is `Convolution` rather than `Kernel` itself that computes the tensor product of `Rs_in` with `Rs_filter` to get `Rs_out`, we do not need Clebsch-Gordan coefficients in `Kernel` forward.

In such a case, the forward einsum would instead read as
```
        kernel = torch.einsum('zk,zk->zk', R, Y)
        return kernel.view(*size, kernel.shape[1], kernel.shape[2])
```

Reply:
Or perhaps, you question is instead -- how do we know that the `tensor_productRs` is giving us the same paths as the loops? If that is the question -- I'm not sure how to prove that without just printing all the paths that the loop version uses and comparing it to the `paths` list that `tensor_productRs` can now output.

Let me know which you mean and I'm happy to explain my thought process further.

Reply:
I realized that my question was not stated clearly enough. I appreciate the effort to guess it’s meaning. It was your second guess. 
Are the path the same? I have the impression that it is the case. 
Is the output matrix the one we really want?

Reply:
My question in pseudocode
```python
# (i) normal usage of tensor_productRs
Rs_out, Q = tensor_productRs(Rs1, Rs2)
out = einsum('kij,i,j->k', in1, in2)

# (ii) usage in Kernel
R_filters, Q = tensor_productRs(Rs_in, Rs_out)
out = einsum('kij,i,k->j', in, filters)
```

- (i) multiply in1 and in2 in all possible ways
- (ii) get all possible path to obtain out starting from in

These two usage are probably valid but it's non trivial for me.

Reply:
Then the next question will concern the normalization in the [line](https://github.com/e3nn/e3nn/blob/master/e3nn/SO3.py#L502)
```python
C = clebsch_gordan(l, l_1, l_2, cached=True) * (2 * l + 1) ** 0.5
```
it should probably depend on the usage (i) or (ii) of the function. (But I'm not sure!)